app/api/utils/AppContext.ts

```diff
@@ -0,0 +1,33 @@
+import { AsyncLocalStorage } from 'async_hooks';
+
+interface ContextData {
+  [k: string]: unknown;
+}
+
+class AppContext {
+  private storage = new AsyncLocalStorage<ContextData>();
+
+  async run(cb: () => Promise<void>, data: ContextData = {}): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.storage.run(data, () => {
+        cb()
+          .then(resolve)
+          .catch(reject);
+      });
+    });
+  }
+
+  get(key: string) {
+    const data = this.storage.getStore();
+    return data && data[key];
+  }
+
+  set(key: string, value: unknown) {
```

Comment:
can we work with strings here instead of unknown ? do you think it will be sufficient for permissions, we should  only need some key information, the user id ?

Reply:
It's probably (still not sure) not enough for permissions. We could use strings and serialize/deserialize the values on the consumer side if needed.
What are your thoughts?

Reply:
then maybe a possible solution is to have a hard typed ContextData and somehow take the specific types belonging to the keys into account on the setter/getter ? im not sure this is possible but having it typed would be preferable.

Reply:
seems, those solutions are not straight forward, maybe leave this as unknown and type guard the accessors of the context with their own type, tenants and permissions know the type and this remains generic with unknown.

Reply:
Yeah, I probably wouldn't hard type the context object and I say that type guarding, or even duck typing is good enough for a key-value storage like this.
If you think its worth we could do something like this:

```
get<T>(key: string): T {
    return this.getContextObject()[key] as T;
  }

  set<T>(key: string, value: T) {
    this.getContextObject()[key] = value;
  }
```

so internally it is still `unknown` and generic, but you could have typings at consumer's level like:

```
appContext.set<string>('key', 'value'); // second param is of type string
const value = appContext.get<string>('key') // return type is string
```

Reply:
Also it's worth mentioning that, unlike `any`, `unknown` is type-safe. That means that ts will force you cast the value to assign it to anything that is not `unknown`.

Reply:
> Also it's worth mentioning that, unlike any, unknown is type-safe. That means that ts will force you cast the value to assign it to anything that is not unknown.

did not know this, thanks really usefull, if ts is going to force us to cast the value then i think this should be enough without the generic types on the setter/getter, sounds good! 👍👍